onnx/defs/tensor/defs.cc

```diff
@@ -313,6 +313,15 @@ ONNX_OPERATOR_SET_SCHEMA(
 
           auto rank = ctx.getInputType(0)->tensor_type().shape().dim_size();
 
+          // rank == 0 may mean that shape inference can't be done yet.
```

Comment:
maybe it's better to check whether ctx.getInputType(0)->tensor_type().has_shape().

This reminds me that a "Status" is still needed to be returned for shape inference function, so that, frameworks calling shape inference in a graph can do early stop (without having to do that via throwing and catching exceptions).

@gramalingam  thoughts?

Reply:
The existing check for “hasNInputShapes(…)” already checks for this. So, I am guessing that the real bug lies elsewhere … it could be in the shape-inference for a preceding node.   

This is somewhat fragile, because of the way “protobuf”’s optional fields work. The “shape” field is optional and not set if it is not known (e.g., if it’s rank is not known). So, every line of code that tries to access “.shape()” should first check if it “.has_shape()”. Otherwise, protobuf accessors will allocate the shape to a default-value when someone accesses it. So, my guess is that some recently added shape-inference code accesses the “.shape()” field without first checking if it “.has_shape()”. 

It is likely to be in the shape-inference for some preceding node that has the same tensor-variable as input or output.

Reply:
As @gramalingam mentioned:  ctx.getInputType(0)->tensor_type().has_shape() is already being checked as part of hasNInputShapes(…)

Reply:
rank() == 0 may mean the tensor is a scalar. And concatenating a scalar with an array is obviously wrong.